File: src/java/org/orbeon/oxf/servlet/OrbeonXFormsFilter.java
/**
 * Copyright (C) 2009 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.servlet;

import javax.servlet.*;
import javax.servlet.http.*;
import java.io.*;
import java.util.*;

/**
 * This filter allows forwarding requests from your web app to an separate Orbeon Forms context.
 *
 * This filter must remain very simple and have as little dependencies as possible, because it must be easily deployable
 * into any web application.
 */
public class OrbeonXFormsFilter implements Filter {

    public static final String RENDERER_DEPLOYMENT_ATTRIBUTE_NAME = "oxf.xforms.renderer.deployment";
    public static final String RENDERER_BASE_URI_ATTRIBUTE_NAME = "oxf.xforms.renderer.base-uri";
    public static final String RENDERER_DOCUMENT_ATTRIBUTE_NAME = "oxf.xforms.renderer.document";
    public static final String RENDERER_CONTENT_TYPE_ATTRIBUTE_NAME = "oxf.xforms.renderer.content-type";
    public static final String RENDERER_HAS_SESSION_ATTRIBUTE_NAME = "oxf.xforms.renderer.has-session";

    public static final String RENDERER_PATH = "/xforms-renderer";

    private static final String RENDERER_CONTEXT_PARAMETER_NAME = "oxf.xforms.renderer.context";
    private static final String DEFAULT_ENCODING = "ISO-8859-1"; // must be this per Servlet spec

    private ServletContext servletContext;
    private String orbeonContextPath;

    public void init(FilterConfig filterConfig) throws ServletException {
        servletContext = filterConfig.getServletContext();
        orbeonContextPath = filterConfig.getInitParameter(RENDERER_CONTEXT_PARAMETER_NAME);

        // TODO: check orbeonContextPath format: starts with /, doesn't end with one, etc.
    }

    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {

        final HttpServletRequest httpRequest = (HttpServletRequest) servletRequest;
        final HttpServletResponse httpResponse = (HttpServletResponse) servletResponse;

        final String requestPath = getRequestPath(httpRequest);

        // Set whether deployment is integrated or separate
        // NOTE: DO this also for resources, so that e.g. /xforms-server, /xforms-server-submit can handle URLs properly
        httpRequest.setAttribute(RENDERER_DEPLOYMENT_ATTRIBUTE_NAME, (getOrbeonContext() == servletContext) ? "integrated" : "separate");

        if (isOrbeonResourceRequest(requestPath)) {
            // Directly forward all requests meant for Orbeon Forms resources (including /xforms-server)
            final String subRequestPath = requestPath.substring(orbeonContextPath.length());
            getOrbeonDispatcher(subRequestPath).forward(httpRequest, httpResponse);
        } else {
            // Forward the request to the Orbeon Forms renderer
            final FilterRequestWrapper requestWrapper = new FilterRequestWrapper(httpRequest);
            final FilterResponseWrapper responseWrapper = new FilterResponseWrapper(httpResponse);

            // Execute filter
            filterChain.doFilter(requestWrapper, responseWrapper);

            // Set document if not present AND output was intercepted
            final boolean isEmptyContent;
            if (httpRequest.getAttribute(RENDERER_DOCUMENT_ATTRIBUTE_NAME) == null) {
                final String content = responseWrapper.getContent();
                if (content != null) {
                    httpRequest.setAttribute(RENDERER_DOCUMENT_ATTRIBUTE_NAME, content);
                }
                isEmptyContent = isBlank(content);
            } else {
                // Assume content is not blank
                isEmptyContent = false;
            }

            // Tell whether there is a session
            httpRequest.setAttribute(RENDERER_HAS_SESSION_ATTRIBUTE_NAME, Boolean.toString(httpRequest.getSession(false) != null));

            // Provide media type if available
            if (responseWrapper.getMediaType() != null)
                httpRequest.setAttribute(RENDERER_CONTENT_TYPE_ATTRIBUTE_NAME, responseWrapper.getMediaType());

            // Set base URI
            httpRequest.setAttribute(RENDERER_BASE_URI_ATTRIBUTE_NAME, requestPath);

            // Forward to Orbeon Forms for rendering only of there is content to be rendered, otherwise just return and
            // let the filterChain finish its life naturally, assuming that when sendRedirect is used, no content is
            // available in the response object
            if (!isEmptyContent) {
                // The request wrapper provides an empty request body if the filtered resource already attempted to
                // read the body.
                final HandleBodyOrbeonRequestWrapper orbeonRequestWrapper
                        = new HandleBodyOrbeonRequestWrapper(httpRequest, requestWrapper.isRequestBodyRead());

                // Forward
                getOrbeonDispatcher(RENDERER_PATH).forward(orbeonRequestWrapper, httpResponse);
            }
        }
    }

    // NOTE: this method copied from Apache StringUtils 2.3 as we don't want a dependency on the JAR
    // http://www.apache.org/licenses/LICENSE-2.0
    public static boolean isBlank(String str) {
        int strLen;
        if (str == null || (strLen = str.length()) == 0) {
            return true;
        }
        for (int i = 0; i < strLen; i++) {
            if ((!Character.isWhitespace(str.charAt(i)))) {
                return false;
            }
        }
        return true;
    }

    public void destroy() {
    }

    private ServletContext getOrbeonContext() throws ServletException {
        final ServletContext orbeonContext = (orbeonContextPath != null) ? servletContext.getContext(orbeonContextPath) : servletContext;
        if (orbeonContext  == null)
            throw new ServletException("Can't find Orbeon Forms context called '" + orbeonContextPath + "'. Check the '" + RENDERER_CONTEXT_PARAMETER_NAME + "' filter initialization parameter and the <Context crossContext=\"true\"/> attribute.");

        return orbeonContext ;
    }

    private RequestDispatcher getOrbeonDispatcher(String path) throws ServletException {
        final RequestDispatcher dispatcher = getOrbeonContext().getRequestDispatcher(path);
        if (dispatcher == null)
            throw new ServletException("Can't find Orbeon Forms request dispatcher.");

        return dispatcher;
    }

    private boolean isOrbeonResourceRequest(String requestPath) {
        return orbeonContextPath != null && requestPath != null && requestPath.startsWith(orbeonContextPath + "/");

    }

    // NOTE: This is borrowed from NetUtils but we don't want the dependency
    private static String getRequestPath(HttpServletRequest request) {

        // Get servlet path and path info
        String servletPath = request.getServletPath();
        if (servletPath == null) servletPath = "";
        String pathInfo = request.getPathInfo();
        if (pathInfo == null) pathInfo = "";

        // Concatenate servlet path and path info, avoiding a double slash
        String requestPath = servletPath.endsWith("/") && pathInfo.startsWith("/")
                ? servletPath + pathInfo.substring(1)
                : servletPath + pathInfo;

        // Add starting slash if missing
        if (!requestPath.startsWith("/"))
            requestPath = "/" + requestPath;

        return requestPath;
    }

    // NOTE: This is borrowed from NetUtils but we don't want the dependency
    public static String getContentTypeCharset(String contentType) {
        if (contentType == null)
            return null;
        int semicolonIndex = contentType.indexOf(";");
        if (semicolonIndex == -1)
            return null;
        int charsetIndex = contentType.indexOf("charset=", semicolonIndex);
        if (charsetIndex == -1)
            return null;
        // FIXME: There may be other attributes after charset, right?
        String afterCharset = contentType.substring(charsetIndex + 8);
        afterCharset = afterCharset.replace('"', ' ');
        return afterCharset.trim();
    }

    // NOTE: This is borrowed from NetUtils but we don't want the dependency
    public static String getContentTypeMediaType(String contentType) {
        if (contentType == null || contentType.equalsIgnoreCase("content/unknown"))
            return null;
        int semicolonIndex = contentType.indexOf(";");
        if (semicolonIndex == -1)
            return contentType;
        return contentType.substring(0, semicolonIndex).trim();
    }

    private static class FilterRequestWrapper extends HttpServletRequestWrapper {

        private boolean requestBodyRead;

        public FilterRequestWrapper(HttpServletRequest httpServletRequest) {
            super(httpServletRequest);

        }

        @Override
        public BufferedReader getReader() throws IOException {
            // If the filtered resource attempts to read the request body, we will forward an empty body
            this.requestBodyRead = true;
            return super.getReader();
        }

        @Override
        public ServletInputStream getInputStream() throws IOException {
            // If the filtered resource attempts to read the request body, we will forward an empty body
            this.requestBodyRead = true;
            return super.getInputStream();
        }

        public boolean isRequestBodyRead() {
            return requestBodyRead;
        }
    }

    private static class HandleBodyOrbeonRequestWrapper extends FilterRequestWrapper {

        // If necessary, return an empty stream for the body, because the body might be read by the JSP and we don't want
        // Orbeon Forms to attempt to read a closed stream.
        private final boolean forceEmptyBody;

        private HandleBodyOrbeonRequestWrapper(HttpServletRequest httpServletRequest, boolean forceEmptyBody) {
            super(httpServletRequest);
            this.forceEmptyBody = forceEmptyBody;
        }

        @Override
        public String getContentType() {
            return forceEmptyBody ? null : super.getContentType();
        }

        @Override
        public int getContentLength() {
            return forceEmptyBody ? 0 : super.getContentLength();
        }

        @Override
        public ServletInputStream getInputStream() throws IOException {
            return forceEmptyBody ? new ServletInputStream() {
                @Override
                public int read() throws IOException {
                    return -1;
                }
            } : super.getInputStream();
        }

        @Override
        public BufferedReader getReader() throws IOException {
            return forceEmptyBody ? new BufferedReader(new Reader() {
                @Override
                public int read(char[] cbuf, int off, int len) throws IOException {
                    return 0;
                }

                @Override
                public void close() throws IOException {
                }
            }) : super.getReader();
        }

        // Filter all headers starting with "if-" so that Orbeon Forms doesn't attempt caching based on those.

        private Enumeration<String> headerNames;

        @Override
        public String getHeader(String s) {
            // Filter conditional get headers so that we always get content
            if (s.toLowerCase().startsWith("if-") )
                return null;
            else
                return super.getHeader(s);
        }

        @Override
        public Enumeration getHeaders(String s) {
            // Filter conditional get headers so that we always get content
            if (s.toLowerCase().startsWith("if-"))
                return null;
            else
                return super.getHeaders(s);
        }

        @Override
        public Enumeration getHeaderNames() {
            if (headerNames == null) {
                // Filter conditional get headers so that we always get content
                final List<String> newHeaderNames = new ArrayList<String>();
                for (Enumeration e = super.getHeaderNames(); e.hasMoreElements();) {
                    final String currentName = (String) e.nextElement();
                    if (!currentName.toLowerCase().startsWith("if-"))
                        newHeaderNames.add(currentName);
                }
                headerNames = Collections.enumeration(newHeaderNames);

            }
            return headerNames;
        }

        @Override
        public long getDateHeader(String s) {
            // Filter conditional get headers so that we always get content
            if (s.toLowerCase().startsWith("if-"))
                return -1;
            else
                return super.getDateHeader(s);
        }
    }

    private static class FilterResponseWrapper extends HttpServletResponseWrapper {

        public FilterResponseWrapper(HttpServletResponse response) {
            super(response);
        }

        private ByteArrayOutputStream byteArrayOutputStream;
        private ServletOutputStream servletOutputStream;

        private StringWriter stringWriter;
        private PrintWriter printWriter;


        private String encoding;
        private String mediatype;

        @Override
        public void addCookie(Cookie cookie) {
            super.addCookie(cookie);
        }

        @Override
        public boolean containsHeader(String string) {
            return super.containsHeader(string);
        }

        @Override
        public String encodeURL(String string) {
            return super.encodeURL(string);
        }

        @Override
        public String encodeRedirectURL(String string) {
            return super.encodeRedirectURL(string);
        }

        @Override
        public String encodeUrl(String string) {
            return super.encodeUrl(string);
        }

        @Override
        public String encodeRedirectUrl(String string) {
            return super.encodeRedirectUrl(string);
        }

        @Override
        public void sendError(int i, String string) throws IOException {
            // TODO
        }

        @Override
        public void sendError(int i) throws IOException {
            // TODO
        }

        @Override
        public void sendRedirect(String string) throws IOException {
            super.sendRedirect(string); 
        }

        @Override
        public void setDateHeader(String string, long l) {
            // TODO
        }

        @Override
        public void addDateHeader(String string, long l) {
            // TODO
        }

        @Override
        public void setHeader(String string, String string1) {
            // TODO
        }

        @Override
        public void addHeader(String string, String string1) {
            // TODO
        }

        @Override
        public void setIntHeader(String string, int i) {
            // TODO
        }

        @Override
        public void addIntHeader(String string, int i) {

            // TODO
        }

        @Override
        public void setStatus(int i) {
            // TODO
        }

        @Override
        public void setStatus(int i, String string) {
            // TODO
        }

        @Override
        public ServletResponse getResponse() {
            return super.getResponse();
        }

        @Override
        public void setResponse(ServletResponse servletResponse) {
            super.setResponse(servletResponse);
        }

        @Override
        public String getCharacterEncoding() {
            // TODO: we don't support setLocale()
            return (encoding == null) ? DEFAULT_ENCODING : encoding;
        }

        public String getMediaType() {
            return mediatype;
        }

        @Override
        public ServletOutputStream getOutputStream() throws IOException {
            if (byteArrayOutputStream == null) {
                byteArrayOutputStream = new ByteArrayOutputStream();
                servletOutputStream = new ServletOutputStream() {
                    public void write(int i) throws IOException {
                        byteArrayOutputStream.write(i);
                    }
                };
            }
            return servletOutputStream;
        }

        @Override
        public PrintWriter getWriter() throws IOException {
            if (printWriter == null) {
                stringWriter = new StringWriter();
                printWriter = new PrintWriter(stringWriter);
            }
            return printWriter;
        }

        @Override
        public void setContentLength(int i) {
            // NOP
        }

        @Override
        public void setContentType(String contentType) {
            this.encoding = getContentTypeCharset(contentType);
            this.mediatype = getContentTypeMediaType(contentType);
        }

        @Override
        public void setBufferSize(int i) {
            // NOP
        }

        @Override
        public int getBufferSize() {
            // We have a buffer, but it is infinite
            return Integer.MAX_VALUE;
        }

        @Override
        public void flushBuffer() throws IOException {
            // NOPE
        }

        @Override
        public boolean isCommitted() {
            // We buffer everything so return false all the time
            return false;
        }

        @Override
        public void reset() {
            resetBuffer();
        }

        @Override
        public void resetBuffer() {
            if (byteArrayOutputStream != null) {
                try {
                    servletOutputStream.flush();
                } catch (IOException e) {
                    // ignore?
                }
                byteArrayOutputStream.reset();
            } else if (stringWriter != null) {
                printWriter.flush();
                final StringBuffer sb = stringWriter.getBuffer();
                sb.delete(0, sb.length());
            }
        }

        @Override
        public void setLocale(Locale locale) {
            // TODO
        }

        @Override
        public Locale getLocale() {
            return super.getLocale();
        }

        public String getContent() throws IOException {
            if (stringWriter != null) {
                // getWriter() was used
                printWriter.flush();
                return stringWriter.toString();
            } else if (servletOutputStream != null) {
                // getOutputStream() was used
                servletOutputStream.flush();
                return new String(byteArrayOutputStream.toByteArray(), getCharacterEncoding());
            } else {
                return null;
            }
        }
    }
}

File: src/main/scala/org/orbeon/oxf/servlet/OrbeonXFormsFilter.scala
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.servlet

import org.orbeon.io.IOUtils._
import org.orbeon.oxf.common.Defaults
import org.orbeon.oxf.http.StatusCode
import org.orbeon.oxf.util.ContentTypes
import org.orbeon.oxf.util.PathUtils._
import org.orbeon.oxf.util.StringUtils._
import org.orbeon.oxf.webapp.ServletSupport

import java.io._
import java.util._


private case class FilterSettings(context: ServletContext, orbeonContextPathOpt: Option[String], defaultEncoding: String) {
  // NOTE: Never match anything if there is no context path
  val OrbeonResourceRegex = orbeonContextPathOpt map (path => s"$path(/.*)".r) getOrElse "$.".r
}

// For backward compatibility
trait OrbeonXFormsFilter extends JavaxOrbeonXFormsFilter

class JavaxOrbeonXFormsFilter   extends JavaxFilter  (new OrbeonXFormsFilterImpl)
class JakartaOrbeonXFormsFilter extends JakartaFilter(new OrbeonXFormsFilterImpl)

// This filter allows forwarding requests from your web app to an separate Orbeon Forms context.
class OrbeonXFormsFilterImpl extends Filter {

  import org.orbeon.oxf.servlet.OrbeonXFormsFilterImpl._

  private var settingsOpt: Option[FilterSettings] = None

  override def init(filterConfig: FilterConfig): Unit =
    settingsOpt =
      Some(
        FilterSettings(
          filterConfig.getServletContext,
          Option(filterConfig.getInitParameter(RendererContextParameterName)) map normalizeContextPath,
          Option(filterConfig.getInitParameter(DefaultEncodingParameterName)) getOrElse Defaults.DefaultEncodingForServletCompatibility
        )
      )

  override def destroy(): Unit = settingsOpt = None

  def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit =
    settingsOpt foreach {
      case settings @ FilterSettings(servletContext, orbeonContextPathOpt, defaultEncoding) =>

        val orbeonContext = (
          orbeonContextPathOpt
          map { orbeonContextPath =>
            servletContext.getContext(orbeonContextPath) ensuring (
              _ ne null,
              s"Can't find Orbeon Forms context called '$orbeonContextPath'. Check the " +
              s"'$RendererContextParameterName' filter initialization parameter and the " +
              s"<Context crossContext='true'/> attribute."
            )
          }
          getOrElse servletContext
        )

        def getOrbeonDispatcher(path: String) =
          orbeonContext.getRequestDispatcher(path) ensuring
          (_ ne null, "Can't find Orbeon Forms request dispatcher.")

        val httpRequest  = servletRequest.asInstanceOf[HttpServletRequest]
        val httpResponse = servletResponse.asInstanceOf[HttpServletResponse]
        val requestPath  = ServletSupport.getRequestPathInfo(httpRequest)

        // Set whether deployment is integrated or separate
        // NOTE: Also for resources so that e.g. /xforms-server, /xforms-server-submit can handle URLs properly.
        httpRequest.setAttribute(
          RendererDeploymentAttributeName,
          if (orbeonContext eq servletContext) "integrated" else "separate"
        )

        requestPath match {
          case settings.OrbeonResourceRegex(subRequestPath) =>
            // Directly forward all requests meant for Orbeon Forms resources (including /xforms-server)

            // Check that the session exists for any request to /xforms-server
            // The purpose of this check is that if the application has invalidated the session, we don't
            // want to allow further interactions with a page.
            // NOTE: With Tomcat this doesn't seem necessary. Other containers might work differently.
            if (subRequestPath.startsWith("/xforms-server"))
              httpRequest.getSession(false) ensuring
              (_ ne null, "Session has expired. Unable to process incoming request.")

            getOrbeonDispatcher(subRequestPath).forward(httpRequest, httpResponse)
          case _ =>
            // Forward the request to the Orbeon Forms renderer
            val requestWrapper  = new FilterRequestWrapper(httpRequest)
            val responseWrapper = new FilterResponseWrapper(httpResponse, defaultEncoding)

            // Execute filter
            filterChain.doFilter(requestWrapper, responseWrapper)

            // Restore content length if needed, see https://github.com/orbeon/orbeon-forms/issues/2775
            //httpResponse.setContentLength(-1)

            val failureCodeOpt =
              responseWrapper.statusCode filterNot StatusCode.isSuccessCode

            failureCodeOpt match {
              case None =>
                // Content can be a String or other content set as a request attribute which can be read by the
                // scope generator
                val content =
                  Option(httpRequest.getAttribute(RendererDocumentAttributeName)) orElse
                  responseWrapper.content

                val nonEmptyContent = content match {
                  case Some(s: String) => s.nonAllBlank
                  case Some(_)         => true
                  case None            => false
                }

                // Forward to Orbeon Forms for rendering only if there is content to be rendered, otherwise just
                // return and let the filterChain finish its life naturally, assuming that when sendRedirect is
                // used, no content is available in the response object.
                if (nonEmptyContent) {

                  // Make sure document is set when available
                  content foreach {
                    httpRequest.setAttribute(
                      RendererDocumentAttributeName,
                      _
                    )
                  }

                  // Whether there is a session
                  httpRequest.setAttribute(
                    RendererHasSessionAttributeName,
                    (httpRequest.getSession(false) ne null).toString
                  )

                  // Mediatype if available
                  responseWrapper.mediaType foreach {
                    httpRequest.setAttribute(
                      RendererContentTypeAttributeName,
                      _
                    )
                  }

                  httpRequest.setAttribute(
                    RendererBaseUriAttributeName,
                    requestPath
                  )

                  // The request wrapper provides an empty request body if the filtered resource already
                  // attempted to read the body.
                  val orbeonRequestWrapper =
                    new OptionalBodyFilterRequestWrapper(httpRequest, requestWrapper.isRequestBodyRead)

                  getOrbeonDispatcher(RendererPath).forward(orbeonRequestWrapper, httpResponse)
                } else {
                  // No content was produced, consider this equivalent to a "not found" status
                  httpResponse.setStatus(404)
                }
              case Some(failureCode) =>
                // Forward status code
                httpResponse.setStatus(failureCode)
            }
        }
    }
}

object OrbeonXFormsFilterImpl {

  val RendererDeploymentAttributeName      = "oxf.xforms.renderer.deployment"
  val RendererBaseUriAttributeName         = "oxf.xforms.renderer.base-uri"
  val RendererDocumentAttributeName        = "oxf.xforms.renderer.document"
  val RendererContentTypeAttributeName     = "oxf.xforms.renderer.content-type"
  val RendererHasSessionAttributeName      = "oxf.xforms.renderer.has-session"

  private val RendererContextParameterName = "oxf.xforms.renderer.context"
  private val DefaultEncodingParameterName = "oxf.xforms.renderer.default-encoding"
  private val RendererPath                 = "/xforms-renderer"

  private def normalizeContextPath(s: String) =
    "/" + s.dropTrailingSlash.dropStartingSlash
}

private class FilterRequestWrapper(httpServletRequest: HttpServletRequest)
    extends HttpServletRequestWrapper(httpServletRequest)
    with RequestRemoveHeaders {

  private var _requestBodyRead = false
  def isRequestBodyRead = _requestBodyRead

  override def getReader = {
    this._requestBodyRead = true
    super.getReader
  }

  override def getInputStream = {
    this._requestBodyRead = true
    super.getInputStream
  }

  // See https://github.com/orbeon/orbeon-forms/issues/1796
  def headersToRemove = (name: String) => name.toLowerCase.startsWith("if-")
}

private class OptionalBodyFilterRequestWrapper(httpServletRequest: HttpServletRequest, forceEmptyBody: Boolean)
    extends FilterRequestWrapper(httpServletRequest) {

  override def getContentType   = if (forceEmptyBody) null else super.getContentType
  override def getContentLength = if (forceEmptyBody) 0    else super.getContentLength

  override def getInputStream =
    if (forceEmptyBody)
      new ServletInputStream {
        def read: Int = -1
        def isFinished: Boolean = true
        def isReady: Boolean = false
        def setReadListener(readListener: ReadListener): Unit = ()
      }
    else
      super.getInputStream

  override def getReader =
    if (forceEmptyBody)
      new BufferedReader(
        new Reader {
          def read(cbuf: Array[Char], off: Int, len: Int) = 0
          def close() = ()
        }
      )
    else
      super.getReader
}

private class FilterResponseWrapper(response: HttpServletResponse, defaultEncoding: String)
    extends HttpServletResponseWrapper(response) {

  private var _statusCode: Option[Int]                      = None
  def statusCode = _statusCode

  private var _byteArrayOutputStream: ByteArrayOutputStream = null
  private var _servletOutputStream: ServletOutputStream     = null
  private var _stringWriter: StringWriter                   = null
  private var _printWriter: PrintWriter                     = null
  private var _encoding: Option[String]                     = None
  private var _mediatype: Option[String]                    = None

  override def setStatus(code: Int)                         = _statusCode = Some(code)
  override def setStatus(code: Int, message: String)        = setStatus(code)
  override def sendError(code: Int, string: String)         = setStatus(code)
  override def sendError(code: Int)                         = setStatus(code)

  override def setDateHeader(name: String, value: Long)     = () // SHOULD HANDLE
  override def addDateHeader(name: String, value: Long)     = () // SHOULD HANDLE
  override def setHeader(name: String, value: String)       = () // SHOULD HANDLE
  override def addHeader(name: String, value: String)       = () // SHOULD HANDLE
  override def setIntHeader(name: String, value: Int)       = () // SHOULD HANDLE
  override def addIntHeader(name: String, value: Int)       = () // SHOULD HANDLE
  override def setLocale(locale: Locale)                    = () // SHOULD HANDLE

  override def setContentLength(i: Int)                     = () // NOP
  override def setBufferSize(i: Int)                        = () // NOP
  override def getBufferSize                                = Integer.MAX_VALUE // as if we had an "infinite" buffer
  override def flushBuffer()                                = () // NOP
  override def isCommitted                                  = false
  override def reset()                                      = resetBuffer()
  override def getCharacterEncoding                         = _encoding getOrElse defaultEncoding

  override def getOutputStream = {
    if (_byteArrayOutputStream eq null) {
      _byteArrayOutputStream = new ByteArrayOutputStream
      _servletOutputStream = new ServletOutputStream {
        def write(i: Int) = _byteArrayOutputStream.write(i)
        def isReady: Boolean = true
        def setWriteListener(writeListener: WriteListener): Unit = ()
      }
    }
    _servletOutputStream
  }

  override def getWriter = {
    if (_printWriter eq null) {
      _stringWriter = new StringWriter
      _printWriter = new PrintWriter(_stringWriter)
    }
    _printWriter
  }

  override def setContentType(contentType: String): Unit = {
    this._encoding  = ContentTypes.getContentTypeCharset(contentType)
    this._mediatype = ContentTypes.getContentTypeMediaType(contentType)
  }

  override def resetBuffer(): Unit = {
    if (_byteArrayOutputStream ne null) {
      runQuietly(_servletOutputStream.flush())
      _byteArrayOutputStream.reset()
    } else if (_stringWriter ne null) {
      _printWriter.flush()
      val sb = _stringWriter.getBuffer
      sb.delete(0, sb.length)
    }
  }

  def mediaType = _mediatype

  def content =
    if (_stringWriter ne null) {
      _printWriter.flush()
      Some(_stringWriter.toString)
    } else if (_servletOutputStream ne null) {
      _servletOutputStream.flush()
      Some(_byteArrayOutputStream.toString(getCharacterEncoding))
    } else
      None
}